src/ast/rewriter/bit_blaster/bit_blaster_tpl_def.h

```diff
@@ -178,175 +157,70 @@ void bit_blaster_tpl<Cfg>::mk_subtracter(unsigned sz, expr * const * a_bits, exp
-                         ...               ...                ...               ...
-                        out[0]            out[1]             out[2]            out[3]
-      
-           HA denotes a half-adder.
-           FA denotes a full-adder.
-        */
-
-        for (unsigned i = 1; i < sz; i++) {
-            checkpoint();
-            couts.reset();
-            expr_ref i1(m()), i2(m());
-            mk_and(a_bits[0], b_bits[i],   i1);
-            mk_and(a_bits[1], b_bits[i-1], i2);
-            if (i < sz - 1) {
-                mk_half_adder(i1, i2, out, cout);
-                couts.push_back(cout);
-                for (unsigned j = 2; j <= i; j++) {
-                    expr_ref prev_out(m());
-                    prev_out = out;
-                    expr_ref i3(m());
-                    mk_and(a_bits[j], b_bits[i-j], i3);
-                    mk_full_adder(i3, prev_out, cins.get(j-2), out, cout);
-                    couts.push_back(cout);
-                }
-                out_bits.push_back(out);
-                cins.swap(couts);
-            }
-            else {
-                // last step --> I don't need to generate/store couts.
-                mk_xor(i1, i2, out);
-                for (unsigned j = 2; j <= i; j++) {
-                    expr_ref i3(m());
-                    mk_and(a_bits[j], b_bits[i-j], i3);
-                    mk_xor3(i3, out, cins.get(j-2), out);
-                }
-                out_bits.push_back(out);
-            }
-        }
-    }
-    else { 
-        // WALLACE TREE MULTIPLIER        
-
-        if (sz == 1) {
-            expr_ref t(m());
-            mk_and(a_bits[0], b_bits[0], t);
-            out_bits.push_back(t);
-            return;
-        }
 
-        // There are sz numbers to add and we use a Wallace tree to reduce that to two. 
-        // In this tree, we reduce as early as possible, as opposed to the Dada tree where some 
-        // additions may be delayed if they don't increase the propagation delay [which may be 
-        // a little bit more efficient, but it's tricky to find out which additions create 
-        // additional delays].
-                
-        expr_ref zero(m());
-        zero = m().mk_false();
-
-        vector< expr_ref_vector > pps;
-        pps.resize(sz, expr_ref_vector(m()));
-               
-        for (unsigned i = 0; i < sz; i++) {
-            checkpoint();
-            // The partial product is a_bits AND b_bits[i] 
-            // [or alternatively ITE(b_bits[i], a_bits, bv0[sz])]
-
-            expr_ref_vector & pp = pps[i];
-            expr_ref t(m());
-            for (unsigned j = 0; j < i; j++)
-                pp.push_back(zero); // left shift by i bits
-            for (unsigned j = 0; j < (sz - i); j++) {
-                mk_and(a_bits[j], b_bits[i], t);
-                pp.push_back(t);
-            }
+    // TODO: share the bool_rewriter with the rest of the bit-blaster
+    bool_rewriter rewriter(m());
```

Comment:
we should really check if setting the parameter "flat" to false helps 
in this case:

```
params_ref p;
p.set_bool("flat", false);
rewriter.updt_params(p);
```